internal/controller/pullrequest_controller.go

```diff
@@ -58,121 +58,101 @@ type PullRequestReconciler struct {
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.17.2/pkg/reconcile
 func (r *PullRequestReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	logger := log.FromContext(ctx)
+	logger.Info("Reconciling PullRequest")
 
 	var pr promoterv1alpha1.PullRequest
-	err := r.Get(ctx, req.NamespacedName, &pr, &client.GetOptions{})
-	if err != nil {
+	if err := r.Get(ctx, req.NamespacedName, &pr); err != nil {
 		if errors.IsNotFound(err) {
 			logger.Info("PullRequest not found", "namespace", req.Namespace, "name", req.Name)
 			return ctrl.Result{}, nil
 		}
-
-		logger.Error(err, "failed to get PullRequest", "namespace", req.Namespace, "name", req.Name)
 		return ctrl.Result{}, fmt.Errorf("failed to get PullRequest: %w", err)
 	}
 
-	pullRequestProvider, err := r.getPullRequestProvider(ctx, pr)
+	provider, err := r.getPullRequestProvider(ctx, pr)
 	if err != nil {
 		return ctrl.Result{}, fmt.Errorf("failed to get PullRequest provider: %w", err)
 	}
-	if pullRequestProvider == nil {
-		return ctrl.Result{}, fmt.Errorf("failed to get pull request provider, pullRequestProvider is nil")
-	}
 
-	deleted, err := r.handleFinalizer(ctx, &pr, pullRequestProvider)
-	if err != nil {
-		return ctrl.Result{}, fmt.Errorf("failed to handle finalizer: %w", err)
+	if deleted, err := r.handleFinalizer(ctx, &pr, provider); err != nil || deleted {
+		return ctrl.Result{}, err
 	}
-	if deleted {
+
+	if pr.Status.State == promoterv1alpha1.PullRequestMerged || pr.Status.State == promoterv1alpha1.PullRequestClosed {
+		logger.Info("Cleaning up close and merged pull request", "pullRequestID", pr.Status.ID)
+		if err := r.Delete(ctx, &pr); err != nil && !errors.IsNotFound(err) {
+			logger.Error(err, "Failed to delete PullRequest")
+			return ctrl.Result{}, fmt.Errorf("failed to delete PullRequest: %w", err)
+		}
 		return ctrl.Result{}, nil
 	}
```

Comment:
Since we do Client.Delete after merge/close, we don't need this logic. We'll hit the finalizer logic instead.